fml/gpu_thread_merger.h

```diff
@@ -0,0 +1,59 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FML_SHELL_COMMON_TASK_RUNNER_MERGER_H_
+#define FML_SHELL_COMMON_TASK_RUNNER_MERGER_H_
+
+#include "flutter/fml/macros.h"
+#include "flutter/fml/memory/ref_counted.h"
+#include "flutter/fml/message_loop_task_queues.h"
+
+namespace fml {
+
+class MessageLoopImpl;
+
+enum class GpuThreadStatus { kRemainsMerged, kRemainsUnmerged, kUnmergedNow };
+
+class GpuThreadMerger : public fml::RefCountedThreadSafe<GpuThreadMerger> {
+ public:
+  // Merges the GPU thread into platform thread for the duration of
+  // the lease term. Lease is managed by the caller by either calling
+  // |ExtendLease| or |DecrementLease|.
+  // When the caller merges with a lease term of say 2. The threads
+  // are going to remain merged until 2 invocations of |DecreaseLease|,
+  // unless an |ExtendLease| gets called.
+  void MergeWithLease(size_t lease_term);
+
+  void ExtendLease(size_t lease_term);
+
+  // Returns true if this call resulted in splitting the
+  // GPU and platform threads. Reduces the lease term by 1.
+  GpuThreadStatus DecrementLease();
+
+  bool IsMerged() const;
+
+  GpuThreadMerger(fml::TaskQueueId platform_queue_id,
+                  fml::TaskQueueId gpu_queue_id);
+
+  // Returns true if the the current thread does *not* own rasterizing.
+  // When the threads are merged, platform thread owns rasterizing.
+  // When un-merged, gpu thread owns rasterizing.
+  bool IsNotOnRasterizingThread();
```

Comment:
uber nit: consider making this `IsOnRasterizingThread` e.g if it ever gets used as `if (!IsNotOnRasterizingThread())` it start to get confusing.

Reply:
Done